test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -1143,16 +1144,88 @@ class test_ucp_sockaddr_wireup_fail : public test_ucp_sockaddr_wireup {
-            UCS_TEST_SKIP_R("don't have p2p lanes to do address pack for");
+        if (fail_wireup_type == FAIL_WIREUP_MSG_SEND) {
+            /* Emulate failure of WIREUP MSG sending by setting the AM Bcopy
+             * function which always return EP_TIMEOUT error */
+            iface->ops.ep_am_bcopy =
+                    reinterpret_cast<uct_ep_am_bcopy_func_t>(
+                            ucs_empty_function_return_bc_ep_timeout);
+        } else if (fail_wireup_type == FAIL_WIREUP_CONNECT_TO_EP) {
+            /* Emulate failure of connecting p2p lanes of peers by setting the
+             * connect_to_ep method to the function that always returns error
+             */
+            iface->ops.ep_connect_to_ep =
+                    reinterpret_cast<uct_ep_connect_to_ep_func_t>(
+                            ucs_empty_function_return_ep_timeout);
+        } else if (fail_wireup_type == FAIL_WIREUP_MSG_ADDR_PACK) {
+            /* Emulate failure of preparation of WIREUP MSG sending by setting
+             * the device address getter to the function that always returns
+             * error */
+            iface->ops.iface_get_device_address =
+                    reinterpret_cast<uct_iface_get_device_address_func_t>(
+                            ucs_empty_function_return_ep_timeout);
         }
     }
 
+    void emulate_failure(entity &e, fail_wireup_t fail_wireup_type)
+    {
+        ucp_worker_h worker = e.worker();
+
+        UCS_ASYNC_BLOCK(&worker->async);
+        if (fail_wireup_type == FAIL_WIREUP_SET_EP_FAILED) {
+            /* Emulate failure of the endpoint by invoking error handling
+             * procedure */
+            ucp_ep_set_failed(e.ep(), UCP_NULL_LANE, UCS_ERR_ENDPOINT_TIMEOUT);
+        } else {
+            if (fail_wireup_type == FAIL_WIREUP_CONNECT_TO_EP) {
+                /* Make sure that stub WIREUP_EP is updated */
+                for (auto lane = 0; lane < ucp_ep_num_lanes(e.ep()); ++lane) {
+                    set_iface_failure(e.ep()->uct_eps[lane]->iface,
+                                      fail_wireup_type);
+                }
+            }
+            for (auto iface_id = 0; iface_id < worker->num_ifaces;
+                 ++iface_id) {
+                set_iface_failure(worker->ifaces[iface_id]->iface,
+                                  fail_wireup_type);
+            }
+        }
+        UCS_ASYNC_UNBLOCK(&worker->async);
+    }
+
+    void check_p2p_lanes(const entity &e, uint32_t wait_ep_flags,
+                         fail_wireup_t fail_wireup_type) const
+    {
+        if (fail_wireup_type == FAIL_WIREUP_CONNECT_TO_EP) {
+            if (!ucp_ep_config(e.ep())->p2p_lanes) {
+                 UCS_TEST_SKIP_R("don't have p2p lanes to do connect for");
+            }
+        } else if (fail_wireup_type == FAIL_WIREUP_MSG_ADDR_PACK) {
+            if (!ucp_ep_config(e.ep())->p2p_lanes &&
+                (wait_ep_flags & UCP_EP_FLAG_SERVER_NOTIFY_CB)) {
```

Comment:
why need to check UCP_EP_FLAG_SERVER_NOTIFY_CB?

Reply:
to make sure this is server side of the connection

Reply:
why does it matter if we skip the test?

Reply:
we should skip the test if no P2P lanes only on server side

Reply:
since REPLY would be sent without including p2p addresses

Reply:
> but shouldn't the p2p lanes situation be same on client and server?

we are sending REPLY message only from server side

Reply:
simplified as we discussed offline